ush/setup.py

```diff
@@ -1804,19 +1633,47 @@ def _dict_find(user_dict, substring):
     # Final failsafe before writing rocoto yaml to ensure we don't have any invalid dicts
     # (e.g. metatasks with no tasks, tasks with no associated commands)
     clean_rocoto_dict(expt_config["rocoto"]["tasks"])
+    expt_config.dereference()
 
-    rocoto_yaml_fp = workflow_config["ROCOTO_YAML_FP"]
-    with open(rocoto_yaml_fp, 'w') as f:
-        yaml.Dumper.ignore_aliases = lambda *args : True
-        yaml.dump(expt_config.get("rocoto"), f, sort_keys=False)
+    rocoto_yaml_fp = Path(workflow_config["ROCOTO_YAML_FP"])
+    rocoto_yaml = get_yaml_config({"workflow": expt_config["rocoto"]})
+    rocoto_yaml.dump(rocoto_yaml_fp)
 
-    var_defns_cfg = get_yaml_config(config=expt_config)
+    var_defns_cfg = get_yaml_config(config=expt_config.data)
     del var_defns_cfg["rocoto"]
 
     # Fixup a couple of data types:
-    for dates in ("DATE_FIRST_CYCL", "DATE_LAST_CYCL"):
-        var_defns_cfg["workflow"][dates] = date_to_str(var_defns_cfg["workflow"][dates])
-    var_defns_cfg.dump(global_var_defns_fp)
+    var_defns_cfg.dump(Path(global_var_defns_fp))
+
+    # Run render on the Rocoto YAML to check for unrendered values.
+    # Quit and report on any found.
+    with StringIO() as buffer:
+        logger = logging.getLogger()
+        handler = logging.StreamHandler(buffer)
+        handler.setLevel(logging.INFO)
+        logger.addHandler(handler)
+        xml_config_str = render(input_file=rocoto_yaml_fp, values_needed=True)
+        values_needed = buffer.getvalue().split("\n")[1:]
+        logger.removeHandler(handler)
+    uwtags = ("!bool", "!float", "!int")
+    not_rendered = any(v for v in values_needed if v.strip() != "jobname") or \
+            any(tag in xml_config_str for tag in uwtags)
+    if not_rendered:
+        # Regex to match '{{' or '{%' but not '{{ jobname }}', as the rocoto
+        # tool adds jobname for each task. Also matches UW-supported tags.
+        pattern = r"({{(?! jobname )|{%.*?%})|!bool|!float|!int"
+        line_not_ok = lambda l: any(m for m in re.finditer(pattern, l))
+        unrendered_lines = [l.strip() for l in xml_config_str.split("\n") if line_not_ok(l)]
+        msg = f"""
+        Jinja expressions remain in the XML configuration file.
+
+        {str(rocoto_yaml_fp)}
+
+        They include:
+
+        {'\n'.join(unrendered_lines)}
+        """
+        raise ValueError(msg)
```

Comment:
Adding a check here to make sure that no Jinja templates exist in the Rocoto XML after being very annoyed by invalid XMLs after while debugging.